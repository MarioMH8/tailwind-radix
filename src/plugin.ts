import type { CSSRuleObject, PluginAPI, PluginCreator, PrefixConfig } from 'tailwindcss/types/config';

import type { Color, ColorName, Palette } from './colors';
import { buildColorName, parseColorName } from './colors';
import type { TailwindRadixColorsOptions } from './options';

/**
 * Generate semantic classes, via Tailwind's `addComponents` API.
 *
 * @see https://tailwindcss.com/docs/plugins#adding-components
 */
function generateSemanticClasses(api: PluginAPI) {
	const palette = api.theme<Palette>('colors');
	const prefix = api.config<PrefixConfig>('prefix');

	for (const [colorName, color] of Object.entries(palette)) {
		if (parseColorName(colorName).dark || !hasAllScales(color)) {
			continue;
		}

		const { darkColorName, foregroundColorName } = findFamily(colorName);

		const darkColor = palette[darkColorName];

		if (!darkColor || !hasAllScales(darkColor)) {
			continue;
		}

		const foregroundColor = palette[foregroundColorName];
		const shouldApplyForeground = foregroundColor && hasAllScales(foregroundColor);

		api.addComponents({
			[`.bg-${colorName}-action`]: apply(
				`bg-${colorName}-4`,
				`hover:bg-${colorName}-5`,
				`active:bg-${colorName}-6`,
				`${prefix}dark:bg-${darkColorName}-4`,
				`${prefix}dark:hover:bg-${darkColorName}-5`,
				`${prefix}dark:active:bg-${darkColorName}-6`
			),
			[`.bg-${colorName}-app`]: apply(`bg-${colorName}-1`, `${prefix}dark:bg-${darkColorName}-1`),
			[`.bg-${colorName}-ghost`]: apply(
				`bg-transparent`,
				`hover:bg-${colorName}-4`,
				`active:bg-${colorName}-5`,
				`${prefix}dark:bg-transparent`,
				`${prefix}dark:hover:bg-${darkColorName}-4`,
				`${prefix}dark:active:bg-${darkColorName}-5`
			),
			[`.bg-${colorName}-solid`]: apply(
				`bg-${colorName}-9`,
				`hover:bg-${colorName}-10`,
				`${prefix}dark:bg-${darkColorName}-9`,
				`${prefix}dark:hover:bg-${darkColorName}-10`,
				shouldApplyForeground ? `text-${foregroundColorName}-12` : ''
			),
			[`.bg-${colorName}-subtle`]: apply(`bg-${colorName}-2`, `${prefix}dark:bg-${darkColorName}-2`),
			[`.bg-${colorName}-ui`]: apply(
				`bg-${colorName}-3`,
				`hover:bg-${colorName}-4`,
				`active:bg-${colorName}-5`,
				`${prefix}dark:bg-${darkColorName}-3`,
				`${prefix}dark:hover:bg-${darkColorName}-4`,
				`${prefix}dark:active:bg-${darkColorName}-5`
			),
			[`.border-${colorName}-dim`]: apply(`border-${colorName}-6`, `${prefix}dark:border-${darkColorName}-6`),
			[`.border-${colorName}-normal`]: apply(
				`border-${colorName}-7`,
				`hover:border-${colorName}-8`,
				`${prefix}dark:border-${darkColorName}-7`,
				`${prefix}dark:hover:border-${darkColorName}-8`
			),
			[`.divide-${colorName}-dim`]: apply(`divide-${colorName}-6`, `${prefix}dark:divide-${darkColorName}-6`),
			[`.divide-${colorName}-normal`]: apply(
				`divide-${colorName}-7`,
				`hover:divide-${colorName}-8`,
				`${prefix}dark:divide-${darkColorName}-7`,
				`${prefix}dark:hover:divide-${darkColorName}-8`
			),
			[`.text-${colorName}-dim`]: apply(`text-${colorName}-11`, `${prefix}dark:text-${darkColorName}-11`),
			[`.text-${colorName}-normal`]: apply(`text-${colorName}-12`, `${prefix}dark:text-${darkColorName}-12`),
		} as CSSRuleObject);
	}
}

/**
 * Check whether a color has all 12 scales needed to generate semantic classes.
 */
function hasAllScales(color: Color) {
	if (typeof color === 'string') {
		return false;
	}

	for (let scale = 1; scale <= 12; scale++) {
		if (!color[scale]) {
			return false;
		}
	}

	return true;
}

/**
 * Each Radix base color name has a corresponding gray scale color name, which,
 * if applied on the foreground against the base color background, can create a
 * more colorful and harmonious vibe.
 *
 * @see https://www.radix-ui.com/colors/docs/palette-composition/composing-a-palette#natural-pairing
 */
const naturalColorPairs: Record<ColorName, ColorName> = {
	amber: 'sand',
	blue: 'slate-dark',
	bronze: 'sand-dark',
	brown: 'sand-dark',
	crimson: 'mauve-dark',
	cyan: 'slate-dark',
	gold: 'sand-dark',
	grass: 'olive-dark',
	gray: 'gray-dark',
	green: 'sage-dark',
	indigo: 'slate-dark',
	iris: 'slate-dark',
	jade: 'sage-dark',
	lime: 'olive',
	mauve: 'mauve-dark',
	mint: 'sage',
	olive: 'olive-dark',
	orange: 'sand-dark',
	pink: 'mauve-dark',
	plum: 'mauve-dark',
	purple: 'mauve-dark',
	red: 'mauve-dark',
	ruby: 'mauve-dark',
	sage: 'sage-dark',
	sand: 'sand-dark',
	sky: 'slate',
	slate: 'slate-dark',
	teal: 'sage-dark',
	tomato: 'mauve-dark',
	violet: 'mauve-dark',
	yellow: 'sand',
} as const;

/**
 * For a given base color name, find its corresponding dark color name and
 * foreground color name.
 */
function findFamily(colorName: string) {
	const { alpha, base, p3 } = parseColorName(colorName);

	// blacka / blackp3a
	if (base === 'black') {
		return {
			darkColorName: buildColorName({
				alpha,
				base: 'white',
				dark: false,
				p3,
			}),
			foregroundColorName: buildColorName({
				alpha: false,
				base: 'white',
				dark: false,
				p3,
			}),
		};
	}

	// whitea / whitep3a
	if (base === 'white') {
		return {
			darkColorName: buildColorName({
				alpha,
				base: 'black',
				dark: false,
				p3,
			}),
			foregroundColorName: buildColorName({
				alpha: false,
				base: 'black',
				dark: false,
				p3,
			}),
		};
	}

	return {
		darkColorName: buildColorName({
			alpha,
			base,
			dark: true,
			p3,
		}),
		foregroundColorName: buildColorName({
			alpha: false,
			base: naturalColorPairs[base] ?? '',
			dark: false,
			p3,
		}),
	};
}

/**
 * Composite utility classes using `@apply`.
 *
 * @see https://github.com/tailwindlabs/tailwindcss/discussions/2049
 */
function apply(...classes: string[]) {
	const processedClasses = classes
		.filter(className => className !== '')
		.map(className => className.replaceAll(' ', '_'));

	return { [`@apply ${processedClasses.join(' ')}`]: {} };
}

/**
 * Build the plugin part of this plugin, which will be the first argument of
 * `plugin.withOptions`.
 *
 * @see https://tailwindcss.com/docs/plugins
 */
export default function createPlugin(options: TailwindRadixColorsOptions = {}): PluginCreator {
	const { disableSemantics } = options;

	if (disableSemantics) {
		return () => {
			/*
			 * Do nothing, since the only purpose of the plugin part is to generate
			 * semantics classes.
			 */
		};
	}

	return generateSemanticClasses;
}
