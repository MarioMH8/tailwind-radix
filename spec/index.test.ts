import { expect, test } from 'bun:test';
import postcss from 'postcss';
import prettier from 'prettier';
import type { Config } from 'tailwindcss';
import tailwind from 'tailwindcss';

import plugin from '../src';
import type { TailwindRadixColorsOptions } from '../src/options';

async function run({
	config = {},
	content,
	options = {},
}: {
	config?: Omit<Config, 'content'>;
	content: string;
	options?: TailwindRadixColorsOptions;
}) {
	const configWithPlugin = {
		...config,
		content: [{ raw: content }],
		plugins: [plugin(options)],
	};

	const result = await postcss(tailwind(configWithPlugin)).process('@tailwind utilities; @tailwind components;', {
		from: undefined,
	});

	return result.css;
}

async function format(source: string) {
	return prettier.format(source, { parser: 'css' });
}

test('Given no option, both utility and semantic classes are generated', async () => {
	const expected = `
    .bg-slate-1 {
      --tw-bg-opacity: 1;
      background-color: rgb(252 252 253 / var(--tw-bg-opacity));
    }
    .bg-slate-app {
      --tw-bg-opacity: 1;
      background-color: rgb(252 252 253 / var(--tw-bg-opacity));
    }
    @media (prefers-color-scheme: dark) {
      .bg-slate-app {
        --tw-bg-opacity: 1;
        background-color: rgb(17 17 19 / var(--tw-bg-opacity));
      }
    }
  `;

	const result = await run({
		content: 'bg-slate-1 bg-slate-app',
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `disableSemantics`, semantic classes are not generated', async () => {
	const expected = `
    .bg-slate-1 {
      --tw-bg-opacity: 1;
      background-color: rgb(252 252 253 / var(--tw-bg-opacity))
    }
  `;

	const result = await run({
		content: 'bg-slate-1 bg-slate-app',
		options: { disableSemantics: true },
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `include`, only specified colors are generated', async () => {
	const expected = `
    .bg-slate-1 {
      --tw-bg-opacity: 1;
      background-color: rgb(252 252 253 / var(--tw-bg-opacity));
    }
  `;

	const result = await run({
		content: 'bg-slate-1 bg-blue-1',
		options: { include: ['slate'] },
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `exclude`, specified colors are not generated', async () => {
	const expected = '';

	const result = await run({
		content: 'bg-slate-1',
		options: { exclude: ['slate'] },
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `priority=radix-only`, Tailwind colors are preserved', async () => {
	const expected = `
    .bg-zinc-100 {
      --tw-bg-opacity: 1;
      background-color: rgb(244 244 245 / var(--tw-bg-opacity));
    }
  `;

	const result = await run({
		content: 'bg-zinc-100',
		options: { priority: 'radix-first' },
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `priority=tailwind-first`, Tailwind colors take precedence', async () => {
	const expected = `
    .bg-red-100 {
      --tw-bg-opacity: 1;
      background-color: rgb(254 226 226 / var(--tw-bg-opacity));
    }
  `;

	const result = await run({
		content: 'bg-red-100',
		options: { priority: 'tailwind-first' },
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Given option `aliases`, conflicted color names are both preserved', async () => {
	const expected = `
    .bg-red-700 {
      --tw-bg-opacity: 1;
      background-color: rgb(185 28 28 / var(--tw-bg-opacity));
    }
    .bg-sun-9 {
      --tw-bg-opacity: 1;
      background-color: rgb(229 72 77 / var(--tw-bg-opacity));
    }
  `;

	const result = await run({
		content: 'bg-sun-9 bg-red-700',
		options: {
			aliases: {
				red: 'sun',
			},
			priority: 'radix-first',
		},
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('Utility classes are generated for custom colors', async () => {
	const expected = `
    .bg-custom1 {
      --tw-bg-opacity: 1;
      background-color: rgb(18 52 86 / var(--tw-bg-opacity));
    }
    .bg-custom2-1 {
      --tw-bg-opacity: 1;
      background-color: rgb(101 67 33 / var(--tw-bg-opacity));
    }
  `;

	const result = await run({
		config: {
			theme: {
				extend: {
					colors: {
						custom1: '#123456',
						custom2: {
							1: '#654321',
						},
					},
				},
			},
		},
		content: 'bg-custom1 bg-custom2-1',
		options: {},
	});

	expect(await format(result)).toEqual(await format(expected));
});

test('CSS classes are in correct shape', async () => {
	const result = await run({
		content:
			'bg-slate-app bg-slate-subtle bg-slate-ui bg-slate-ghost bg-slate-action border-slate-dim border-slate-normal divide-slate-dim divide-slate-normal text-slate-dim text-slate-normal',
	});

	expect(await format(result)).toMatchSnapshot();
});
