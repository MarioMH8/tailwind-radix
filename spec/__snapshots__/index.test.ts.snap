// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`CSS classes are in correct shape 1`] = `
".bg-slate-action {
  --tw-bg-opacity: 1;
  background-color: rgb(232 232 236 / var(--tw-bg-opacity));
}
.bg-slate-action:hover {
  --tw-bg-opacity: 1;
  background-color: rgb(224 225 230 / var(--tw-bg-opacity));
}
.bg-slate-action:active {
  --tw-bg-opacity: 1;
  background-color: rgb(217 217 224 / var(--tw-bg-opacity));
}
@media (prefers-color-scheme: dark) {
  .bg-slate-action {
    --tw-bg-opacity: 1;
    background-color: rgb(39 42 45 / var(--tw-bg-opacity));
  }
  .bg-slate-action:hover {
    --tw-bg-opacity: 1;
    background-color: rgb(46 49 53 / var(--tw-bg-opacity));
  }
  .bg-slate-action:active {
    --tw-bg-opacity: 1;
    background-color: rgb(54 58 63 / var(--tw-bg-opacity));
  }
}
.bg-slate-app {
  --tw-bg-opacity: 1;
  background-color: rgb(252 252 253 / var(--tw-bg-opacity));
}
@media (prefers-color-scheme: dark) {
  .bg-slate-app {
    --tw-bg-opacity: 1;
    background-color: rgb(17 17 19 / var(--tw-bg-opacity));
  }
}
.bg-slate-ghost {
  background-color: transparent;
}
.bg-slate-ghost:hover {
  --tw-bg-opacity: 1;
  background-color: rgb(232 232 236 / var(--tw-bg-opacity));
}
.bg-slate-ghost:active {
  --tw-bg-opacity: 1;
  background-color: rgb(224 225 230 / var(--tw-bg-opacity));
}
@media (prefers-color-scheme: dark) {
  .bg-slate-ghost {
    background-color: transparent;
  }
  .bg-slate-ghost:hover {
    --tw-bg-opacity: 1;
    background-color: rgb(39 42 45 / var(--tw-bg-opacity));
  }
  .bg-slate-ghost:active {
    --tw-bg-opacity: 1;
    background-color: rgb(46 49 53 / var(--tw-bg-opacity));
  }
}
.bg-slate-subtle {
  --tw-bg-opacity: 1;
  background-color: rgb(249 249 251 / var(--tw-bg-opacity));
}
@media (prefers-color-scheme: dark) {
  .bg-slate-subtle {
    --tw-bg-opacity: 1;
    background-color: rgb(24 25 27 / var(--tw-bg-opacity));
  }
}
.bg-slate-ui {
  --tw-bg-opacity: 1;
  background-color: rgb(240 240 243 / var(--tw-bg-opacity));
}
.bg-slate-ui:hover {
  --tw-bg-opacity: 1;
  background-color: rgb(232 232 236 / var(--tw-bg-opacity));
}
.bg-slate-ui:active {
  --tw-bg-opacity: 1;
  background-color: rgb(224 225 230 / var(--tw-bg-opacity));
}
@media (prefers-color-scheme: dark) {
  .bg-slate-ui {
    --tw-bg-opacity: 1;
    background-color: rgb(33 34 37 / var(--tw-bg-opacity));
  }
  .bg-slate-ui:hover {
    --tw-bg-opacity: 1;
    background-color: rgb(39 42 45 / var(--tw-bg-opacity));
  }
  .bg-slate-ui:active {
    --tw-bg-opacity: 1;
    background-color: rgb(46 49 53 / var(--tw-bg-opacity));
  }
}
.border-slate-dim {
  --tw-border-opacity: 1;
  border-color: rgb(217 217 224 / var(--tw-border-opacity));
}
@media (prefers-color-scheme: dark) {
  .border-slate-dim {
    --tw-border-opacity: 1;
    border-color: rgb(54 58 63 / var(--tw-border-opacity));
  }
}
.border-slate-normal {
  --tw-border-opacity: 1;
  border-color: rgb(205 206 214 / var(--tw-border-opacity));
}
.border-slate-normal:hover {
  --tw-border-opacity: 1;
  border-color: rgb(185 187 198 / var(--tw-border-opacity));
}
@media (prefers-color-scheme: dark) {
  .border-slate-normal {
    --tw-border-opacity: 1;
    border-color: rgb(67 72 78 / var(--tw-border-opacity));
  }
  .border-slate-normal:hover {
    --tw-border-opacity: 1;
    border-color: rgb(90 97 105 / var(--tw-border-opacity));
  }
}
.divide-slate-dim > :not([hidden]) ~ :not([hidden]) {
  --tw-divide-opacity: 1;
  border-color: rgb(217 217 224 / var(--tw-divide-opacity));
}
@media (prefers-color-scheme: dark) {
  .divide-slate-dim > :not([hidden]) ~ :not([hidden]) {
    --tw-divide-opacity: 1;
    border-color: rgb(54 58 63 / var(--tw-divide-opacity));
  }
}
.divide-slate-normal > :not([hidden]) ~ :not([hidden]) {
  --tw-divide-opacity: 1;
  border-color: rgb(205 206 214 / var(--tw-divide-opacity));
}
.divide-slate-normal:hover > :not([hidden]) ~ :not([hidden]) {
  --tw-divide-opacity: 1;
  border-color: rgb(185 187 198 / var(--tw-divide-opacity));
}
@media (prefers-color-scheme: dark) {
  .divide-slate-normal > :not([hidden]) ~ :not([hidden]) {
    --tw-divide-opacity: 1;
    border-color: rgb(67 72 78 / var(--tw-divide-opacity));
  }
  .divide-slate-normal:hover > :not([hidden]) ~ :not([hidden]) {
    --tw-divide-opacity: 1;
    border-color: rgb(90 97 105 / var(--tw-divide-opacity));
  }
}
.text-slate-dim {
  --tw-text-opacity: 1;
  color: rgb(96 100 108 / var(--tw-text-opacity));
}
@media (prefers-color-scheme: dark) {
  .text-slate-dim {
    --tw-text-opacity: 1;
    color: rgb(176 180 186 / var(--tw-text-opacity));
  }
}
.text-slate-normal {
  --tw-text-opacity: 1;
  color: rgb(28 32 36 / var(--tw-text-opacity));
}
@media (prefers-color-scheme: dark) {
  .text-slate-normal {
    --tw-text-opacity: 1;
    color: rgb(237 238 240 / var(--tw-text-opacity));
  }
}
"
`;
